test/SILOptimizer/definite_init_closures.swift

```diff
@@ -0,0 +1,50 @@
+// RUN: %target-swift-frontend -emit-sil %s -o /dev/null
+
+// Test boolean operators with implicit closures
+
+struct Simple {
+  let x: Bool
+  let y: Bool
+
+  init() {
+    x = false
+    y = false || x
+  }
+}
+
+struct NestedClosures {
+  let x: Bool
+  let y: Bool
+  let z: Bool
+
+  init(_ a: Bool) {
+    x = false
+    y = false
+    z = false || (y || (x || a))
+  }
+}
+
+class SimpleKlass {
+  let x: Bool
+  let y: Bool
+
+  init() {
+    x = false
+    y = false || x
+  }
+}
+
+func forward(_ b: inout Bool) -> Bool {
+  return b
+}
+
+struct InoutUse {
+  var x: Bool
+  var y: Bool
+
+  init() {
+    x = false
+    y = false || forward(&x)
```

Comment:
Can you add some invalid test cases as well, to ensure we still diagnose when the field is not initialized?

Reply:
Another case to test is partial initialization:

```
init(b: Bool) {
  if b {
    x = false
  }

  y = false || x
}
```

Reply:
The invalid  test cases are in `definite_init_closures_fail.swift`.
I added a test for partial initialization